broadcaster/_backends/pulsar.py

```diff
@@ -14,64 +15,80 @@ def __init__(self, url: str):
             raise e
 
     async def disconnect(self) -> None:
-        if self._producer:
-            await anyio.to_thread.run_sync(self._producer.close)
-        if self._consumer:
-            await anyio.to_thread.run_sync(self._consumer.close)
+        for producer in self._producers.values():
+            await anyio.to_thread.run_sync(producer.close)
+        for consumer in self._consumers.values():
+            await anyio.to_thread.run_sync(consumer.close)
         if self._client:
             await anyio.to_thread.run_sync(self._client.close)
 
     async def subscribe(self, channel: str) -> None:
-        # In this implementation, we're using a single topic 'broadcast'
-        # So we don't need to do anything here
-        pass
+        if channel not in self._subscribed_channels:
+            self._subscribed_channels.add(channel)
+            consumer = await anyio.to_thread.run_sync(
+                lambda: self._client.subscribe(
+                    channel,
+                    subscription_name=f"broadcast_subscription_{channel}",
+                    consumer_type=pulsar.ConsumerType.Shared,
+                )
+            )
+            self._consumers[channel] = consumer
+            logger.info(f"Subscribed to channel: {channel}")
 
     async def unsubscribe(self, channel: str) -> None:
-        # Similarly, we don't need to do anything here
-        pass
+        if channel in self._subscribed_channels:
+            self._subscribed_channels.remove(channel)
+            consumer = self._consumers.pop(channel, None)
+            if consumer:
+                await anyio.to_thread.run_sync(consumer.close)
+            logger.info(f"Unsubscribed from channel: {channel}")
 
     async def publish(self, channel: str, message: typing.Any) -> None:
-        encoded_message = f"{channel}:{message}".encode("utf-8")
-        await anyio.to_thread.run_sync(lambda: self._producer.send(encoded_message))
+        if channel not in self._producers:
+            self._producers[channel] = await anyio.to_thread.run_sync(
+                lambda: self._client.create_producer(channel)
+            )
+        encoded_message = str(message).encode("utf-8")
+        await anyio.to_thread.run_sync(lambda: self._producers[channel].send(encoded_message))
+        logger.info(f"Published message to channel {channel}: {message}")
 
     async def next_published(self) -> Event:
         while True:
-            try:
-                msg = await anyio.to_thread.run_sync(self._consumer.receive)
-                channel, content = msg.data().decode("utf-8").split(":", 1)
-                await anyio.to_thread.run_sync(lambda: self._consumer.acknowledge(msg))
-                return Event(channel=channel, message=content)
-
-            except anyio.get_cancelled_exc_class():
-                # cancellation
-                logging.info("next_published task is being cancelled")
-                raise
+            if not self._consumers:
```

Comment:
Better to create a background task for each consumer that receives events from a specific channel, then make them all push into a shared queue and make the `next_published` get from that shared queue.
This way, receivers are tied to consumers and the next published get simplified.

Make sure to limit the queue size (to make the receivers wait on pushing) and terminate receivers on disconnect / unsubscribe.